pytest_perf/runner.py

```diff
@@ -74,35 +75,89 @@ class BenchmarkRunner:
         return pip_run.launch._setup_env(target)
 
-    def run(self, cmd: Command):
-        experiment = self.eval(cmd, env=self.experiment_env)
-        control = self.eval(cmd, env=self.control_env)
-        return Result(control, experiment)
+    def run(self, cmd: Command) -> Result:
+        try:
+            experiment = self.eval(cmd, env=self.experiment_env)
+            control = self.eval(cmd, env=self.control_env)
+            return Result(control, experiment)
+        finally:
+            self.stack.close()
 
-    def eval(self, cmd, **kwargs):
+    def eval(self, cmd: List[str], **kwargs) -> str:
         with tempfile.TemporaryDirectory() as empty:
-            out = subprocess.check_output(cmd, cwd=empty, **_text, **kwargs)
-        val = re.search(r'([0-9.]+ \w+) per loop', out).group(1)
+            out = subprocess.check_output(  # type: ignore
+                cmd, cwd=empty, **_text, **kwargs
+            )
+        val = re.search(r'([0-9.]+ \w+) per loop', out).group(1)  # type: ignore
         return val
 
 
-def upstream_url(extras='', control=None):
+def upstream_url() -> str:
     """
     >>> upstream_url()
-    'pytest-perf@git+https://github.com/jaraco/pytest-perf'
+    'https://github.com/jaraco/pytest-perf'
     """
     cmd = ['git', 'remote', 'get-url', 'origin']
-    origin = subprocess.check_output(cmd, **_text).strip()
-    base, sep, name = origin.rpartition('/')
-    rev = f'@{control}' if control else ''
-    return f'{name}{extras}@git+{origin}{rev}'
+    return subprocess.check_output(cmd, **_text).strip()  # type: ignore
+
+
+@contextlib.contextmanager
+def upstream_package(
+    url: Optional[str] = None, control: Optional[str] = None
+) -> Iterator[str]:
+    """Clone the upstream URL to be used as a local package.
+    This avoids the need of knowing the package name to install optional dependencies.
+    """
+    url = url or upstream_url()
+    rev: List[str] = ['--branch', control] if control else []
+    with tempfile.TemporaryDirectory() as tmp:
+        cmd = ['git', 'clone', '--depth', '1', *rev, url, str(tmp)]
+        subprocess.run(cmd, check=True, **_text)  # type: ignore
+        with local_package(tmp) as target:
+            yield target
+
+            if os.name == "nt":
+                # Avoid cleanup errors on Windows
+                subprocess.run("rmdir /S /Q .git", shell=True, check=False)
+
+
+@contextlib.contextmanager
+def local_package(target: str) -> Iterator[str]:
+    """Installation context for local package to workaround ``pip`` limitations
+    (``pip`` only allow extras for a distribution archive or ``.``,
+    not local directories in general).
+
+    The installation context will ``chdir`` into an directory from where the
+    installation should take place and bind the name of the installable item
+    to the target of the ``with ... as`` clause.
+    """
+    if target != "." and os.path.isdir(target):
+        _orig_dir = os.getcwd()
+        os.chdir(target)
```

Comment:
I'm uneasy about this approach. A chdir operation is thread-unsafe because it affects the whole process. I think it should be possible instead have the installer run in that directory.